packages/flutter_tools/lib/src/_flutter_manifest.dart

```diff
@@ -0,0 +1,460 @@
+// found in the LICENSE file.
+
+import 'package:pub_semver/pub_semver.dart';
+import 'package:yaml/yaml.dart';
+
+import 'base/file_system.dart';
+import 'base/user_messages.dart';
+import 'base/utils.dart';
+import 'globals.dart';
+
+/// The type of the Flutter project created using `flutter create -t <type>`
+enum ProjectType {
+  /// A Flutter project is considered an app when it doesn't a `module:`
+  /// or `plugin:` descriptor.
+  ///
+  /// Such a project can be created using `flutter create -t app`.
+  app,
+
+  /// A Flutter project is considered a module when it has a `module:`
+  /// descriptor. A Flutter module project supports integration into an
+  /// existing host app, and has managed platform host code.
+  ///
+  /// Such a project can be created using `flutter create -t module`.
+  module,
+
+  /// A Flutter project is considered a plugin when it has a `plugin:`
+  /// descriptor. A Flutter plugin project wraps custom Android and/or
+  /// iOS code in a Dart interface for consumption by other Flutter app
+  /// projects.
+  ///
+  /// Such a project can be created using `flutter create -t plugin`.
+  plugin,
+}
+
+/// A wrapper around the `flutter:` section in the `pubspec.yaml` file.
+final class FlutterManifest {
+  FlutterManifest({
+    this.appName = '',
+    this.appVersion = '',
+    this.buildName = '',
+    this.buildNumber = '',
+    this.usesMaterialDesign = false,
+    this.projectType = ProjectType.app,
+    this.assets = const <Uri>[],
+    this.fonts = const <Font>[],
+  });
+
+  /// The string value of the top-level `name` property in the `pubspec.yaml` file.
+  final String appName;
+
+  /// The version String from the `pubspec.yaml` file.
+  /// Can be null if it isn't set or has a wrong format.
+  final String appVersion;
+
+  /// The build version name from the `pubspec.yaml` file.
+  /// Can be null if version isn't set or has a wrong format.
+  final String buildName;
+
+  /// The build version number from the `pubspec.yaml` file.
+  /// Can be null if version isn't set or has a wrong format.
+  final String buildNumber;
+
+  /// True if the project uses Material Design.
+  final bool usesMaterialDesign;
+
+  /// The type of Flutter project.
+  final ProjectType projectType;
+
+  /// The list of assets.
+  final List<Uri> assets;
+
+  /// The list of fonts.
+  final List<Font> fonts;
+
+  @override
+  String toString() {
+    return '$runtimeType(appName: $appName, appVersion: $appVersion, buildName: $buildName, '
+        'buildNumber: $buildNumber, projectType: $projectType)';
+  }
+}
+
+/// A wrapper around each of the sections under `fonts:` in `pubspec.yaml`.
+final class Font {
+  Font(this.familyName, this.fontAssets)
+    : assert(familyName != null),
+      assert(fontAssets != null),
+      assert(fontAssets.isNotEmpty);
+
+  final String familyName;
+  final List<FontAsset> fontAssets;
+
+  Map<String, dynamic> get descriptor {
```

Comment:
Prefer using `Object` over `dynamic`. They are both more or less than type, `dynamic` is just an `Object` that allows you to call methods without casting (the checks are performed at runtime). This is more trouble than it is worth in most cases